### PR TITLE
[core] Fixing bug when dispatching tasks to deleted placement group

### LINKED ISSUE #14092: [core] removing placement group with a scheduled actor raises "!is_infeasible" and crashes Ray Tune
### What is the problem?

Ray version: Latest release (1.2.0.), latest wheels
Mac OS

Removing a placement group with a just scheduled actor on it throws an `is_infeasible` error. Interestingly, this only happens when the placement group is re-used multiple times.

In Ray Tune, this leads to a SIGABRT/code 134 error when running a test where Tune dies right after scheduling a trial on a re-used placement group. When killing the placement groups (either manually or by starting a new tune run), sometimes the error is printed, but the script is always completely aborted. It is not clear to me why the simple reproduction script below does not throw SIGABRT. A repro script using Ray Tune with placement groups is also given below.

cc @wuisawesome @rkooo567 

### Reproduction (REQUIRED)
```
import time

import ray
from ray.util import placement_group, remove_placement_group


@ray.remote
class Actor:
    def train(self, s=1):
        time.sleep(s)
        return 1


def test_leak():
    ray.init(num_cpus=2)

    pg = placement_group([{"CPU": 2}], name="test")
    not_ready = [pg.ready()]
    while not_ready:
        _, not_ready = ray.wait(not_ready, timeout=0.1)

    for i in range(2):  # Only works when this is >= 2
        actor = Actor.options(placement_group=pg, placement_group_bundle_index=0, num_cpus=2).remote()
        fut = actor.train.remote(1)
        ray.get(fut)
        actor.__ray_terminate__.remote()

    actor = Actor.options(placement_group=pg, placement_group_bundle_index=0, num_cpus=2).remote()
    fut = actor.train.remote(10)

    remove_placement_group(pg)

    time.sleep(5)


if __name__ == "__main__":
    test_leak()

```

Run with `pytest -s leak.py`:

```
(raylet) [2021-02-14 00:06:46,764 C 66738 8903793] cluster_task_manager.cc:260:  Check failed: !is_infeasible Task cannot be infeasible when it is about to be dispatched
(raylet) [2021-02-14 00:06:46,764 E 66738 8903793] logging.cc:435: *** Aborted at 1613257606 (unix time) try "date -d @1613257606" if you are using GNU date ***
(raylet) [2021-02-14 00:06:46,764 E 66738 8903793] logging.cc:435: PC: @                0x0 (unknown)
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435: *** SIGABRT (@0x7fff71a5533a) received by PID 66738 (TID 0x10a910dc0) stack trace: ***
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @     0x7fff71b065fd _sigtramp
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @     0x7fecc88aff30 (unknown)
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @     0x7fff719dc808 abort
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102f654bc ray::SpdLogMessage::Flush()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102f39f89 ray::RayLog::~RayLog()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102b73698 ray::raylet::ClusterTaskManager::AttemptDispatchWork()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102b724bd ray::raylet::ClusterTaskManager::DispatchScheduledTasksToWorkers()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102aec178 ray::raylet::NodeManager::HandleWorkerAvailable()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102ae6df2 ray::raylet::NodeManager::HandleWorkerAvailable()
(raylet) [2021-02-14 00:06:46,765 E 66738 8903793] logging.cc:435:     @        0x102ae6cc0 ray::raylet::NodeManager::ProcessAnnounceWorkerPortMessage()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x102ae551a ray::raylet::NodeManager::ProcessClientMessage()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x102b545dc std::__1::__function::__func<>::operator()()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x102f0f6fe ray::ClientConnection::ProcessMessage()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x102f17420 boost::asio::detail::reactive_socket_recv_op<>::do_complete()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x1033d9593 boost::asio::detail::scheduler::do_run_one()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x1033cd3d2 boost::asio::detail::scheduler::run()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x1033cd26b boost::asio::io_context::run()
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @        0x102a9eed8 main
(raylet) [2021-02-14 00:06:46,766 E 66738 8903793] logging.cc:435:     @     0x7fff7190dcc9 start
```

Reproduction on Ray Tune (script exits with SIGABRT/code 134):

1. Checkout https://github.com/ray-project/ray/pull/13906
2. Run `pytest -s python/ray/tune/tests/test_tune_restore.py::TuneFailResumeGridTest::testFailResumeAfterPreset`

Removal of placement groups is triggered when starting the second ray tune run (which kills the old existing placement groups)

Example output on CI: https://travis-ci.com/github/ray-project/ray/jobs/482603352#L794 (spoiler: empty)
Output on my local machine:

```
(raylet) [2021-02-14 00:15:14,717 C 67303 8913631] cluster_task_manager.cc:260:  Check failed: !is_infeasible Task cannot be infeasible when it is about to be dispatched
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435: *** Aborted at 1613258114 (unix time) try "date -d @1613258114" if you are using GNU date ***
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435: PC: @                0x0 (unknown)
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435: *** SIGABRT (@0x7fff71a5533a) received by PID 67303 (TID 0x10ad3fdc0) stack trace: ***
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435:     @     0x7fff71b065fd _sigtramp
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435:     @     0x7fe8c5e05000 (unknown)
(raylet) [2021-02-14 00:15:14,717 E 67303 8913631] logging.cc:435:     @     0x7fff719dc808 abort
(raylet) [2021-02-14 00:15:14,722 E 67303 8913631] logging.cc:435:     @        0x10434e4bc ray::SpdLogMessage::Flush()
(raylet) [2021-02-14 00:15:14,723 E 67303 8913631] logging.cc:435:     @        0x104322f89 ray::RayLog::~RayLog()
(raylet) [2021-02-14 00:15:14,723 E 67303 8913631] logging.cc:435:     @        0x103f5c698 ray::raylet::ClusterTaskManager::AttemptDispatchWork()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103f5b4bd ray::raylet::ClusterTaskManager::DispatchScheduledTasksToWorkers()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103ed5178 ray::raylet::NodeManager::HandleWorkerAvailable()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103ecfdf2 ray::raylet::NodeManager::HandleWorkerAvailable()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103ecfcc0 ray::raylet::NodeManager::ProcessAnnounceWorkerPortMessage()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103ece51a ray::raylet::NodeManager::ProcessClientMessage()
(raylet) [2021-02-14 00:15:14,724 E 67303 8913631] logging.cc:435:     @        0x103f3d5dc std::__1::__function::__func<>::operator()()
(raylet) [2021-02-14 00:15:14,725 E 67303 8913631] logging.cc:435:     @        0x1042f86fe ray::ClientConnection::ProcessMessage()
(raylet) [2021-02-14 00:15:14,725 E 67303 8913631] logging.cc:435:     @        0x104300420 boost::asio::detail::reactive_socket_recv_op<>::do_complete()
(raylet) [2021-02-14 00:15:14,726 E 67303 8913631] logging.cc:435:     @        0x1047c2593 boost::asio::detail::scheduler::do_run_one()
(raylet) [2021-02-14 00:15:14,726 E 67303 8913631] logging.cc:435:     @        0x1047b63d2 boost::asio::detail::scheduler::run()
(raylet) [2021-02-14 00:15:14,726 E 67303 8913631] logging.cc:435:     @        0x1047b626b boost::asio::io_context::run()
(raylet) [2021-02-14 00:15:14,726 E 67303 8913631] logging.cc:435:     @        0x103e87ed8 main
(raylet) [2021-02-14 00:15:14,726 E 67303 8913631] logging.cc:435:     @     0x7fff7190dcc9 start
zsh: abort      pytest -s 
```

PyCharm:

```
Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
When doing dispatch, some tasks might be infeasible to run because the placement group got deleted. Put these tasks back to the infeasible queue to fix this problem and prevent the worker from crashing.

## Related issue number
Closes #14092

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_placement_group.py [hunk 1]
```diff
@@ -370,6 +370,38 @@ def f():
         ray.get(task_ref)
 
 
+@pytest.mark.parametrize(
+    "ray_start_cluster_head", [{
+        "num_cpus": 2
+    }], indirect=True)
+def test_remove_placement_group2(ray_start_cluster_head):
```
Comment (iycheng) [inline python/ray/tests/test_placement_group.py:377]: This test is not crashing even before this PR. I guess the underline worker crashing doesn't affect the driver from running successfully. Any suggestion for the change?

Comment (richardliaw) [reply python/ray/tests/test_placement_group.py:377]: One possible way of testing this is to remove https://github.com/ray-project/ray/blob/master/python/ray/tune/tests/test_tune_restore.py#L209-L216

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 2]
```diff
@@ -248,28 +249,26 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
       }
       work_it = dispatch_queue.erase(work_it);
     }
-    if (dispatch_queue.empty()) {
+    if (is_infeasible) {
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:252]: Could we add a unit test for this?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
+1 to Eric's comment about a test case. Probably a c++ unit test at the minimum (probably something that looks like `ClusterTaskManagerTest::ResourceTakenWhileResolving`

File: src/ray/raylet/scheduling/cluster_task_manager_test.cc [hunk 3]
```diff
@@ -284,8 +288,8 @@ TEST_F(ClusterTaskManagerTest, ResourceTakenWhileResolving) {
       std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234);
   std::shared_ptr<MockWorker> worker2 =
       std::make_shared<MockWorker>(WorkerID::FromRandom(), 12345);
-  pool_.PushWorker(std::dynamic_pointer_cast<WorkerInterface>(worker2));
-  pool_.PushWorker(std::dynamic_pointer_cast<WorkerInterface>(worker));
+  pool_.PushWorker(std::static_pointer_cast<WorkerInterface>(worker2));
```
Comment (iycheng) [inline src/ray/raylet/scheduling/cluster_task_manager_test.cc:291]: It's irrelevant with the PR and hope it's not annoying. 
`static_cast` here is enough and I feel it's not worth putting it in another PR.

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
@iycheng change the tag to "tests-ok" if you think this is ready to merge

**Pull request merged**